### PR TITLE
introduce `USE_LOCAL_ARTIFACTS` for speeding up development

### PR DESCRIPTION
**Pull request by AaronKutch**
I should probably add a note in one of the READMEs, but I'm not sure which one

### CODE REVIEW

File: tests/build-container.sh [hunk 1]
```diff
@@ -6,11 +6,49 @@ DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
...
+else
+    # getting the `test-runner` binary
+    pushd $REPOFOLDER/orchestrator/test_runner && PATH=$PATH:$HOME/.cargo/bin cargo build --all --release
+    # getting the `gravity` binary. `GOBIN` is set so that it is placed under `/dockerfile`.
+    # We check for its existence and rename it to prevent confusion.
+    # This will be moved to its normal place by the `Dockerfile`.
+    pushd $REPOFOLDER/module/ &&
+        PATH=$PATH:/usr/local/go/bin GOPROXY=https://proxy.golang.org make &&
+        PATH=$PATH:/usr/local/go/bin GOBIN=$REPOFOLDER/tests/dockerfile make install &&
+        mv $REPOFOLDER/tests/dockerfile/gravity $REPOFOLDER/tests/dockerfile/go_gravity_bin
```
Comment (dzmitryhil) [inline tests/build-container.sh:29]: let's keep the original name after the copy. "go_gravity_bin"  -> "gravity"

File: tests/build-container.sh [hunk 2]
```diff
@@ -6,11 +6,49 @@ DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
...
+
+    # TODO I would like to also use the local npm artifacts,
+    # but it seems to require more than just copying the binaries
+    # get the ts-node and typescript binaries.
+    #pushd $REPOFOLDER/solidity/ &&
```
Comment (dzmitryhil) [inline tests/build-container.sh:34]: What you can do here is to copy "node_modules" at least, it would improve the speed as well

File: tests/container-scripts/integration-tests.sh [hunk 3]
```diff
@@ -16,4 +16,9 @@ killall -9 test-runner
 set -e
 
 pushd /gravity/orchestrator/test_runner
-RUST_BACKTRACE=full TEST_TYPE=$TEST_TYPE RUST_LOG=INFO PATH=$PATH:$HOME/.cargo/bin cargo run --release --bin test-runner
+
+if [[ "${USE_LOCAL_ARTIFACTS:-0}" -eq "0" ]]; then
+    RUST_BACKTRACE=full TEST_TYPE=$TEST_TYPE RUST_LOG=INFO PATH=$PATH:$HOME/.cargo/bin cargo run --release --bin test-runner
```
Comment (dzmitryhil) [inline tests/container-scripts/integration-tests.sh:21]: you can create a variables like "RUN_ARGS" with the "--release --bin test-runner" or "/gravity/orchestrator/target/release/test-runner" depends on "USE_LOCAL_ARTIFACTS". Same for the next script. To reduce the code duplication.

File: tests/build-container.sh [hunk 4]
```diff
@@ -6,11 +6,49 @@ DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
...
+    #    # installs in $REPOFOLDER/solidity/node_modules/bin/ts-node
+    #    npm install --prefix $REPOFOLDER/solidity/node_modules -g ts-node &&
+    #    npm install --prefix $REPOFOLDER/solidity/node_modules -g typescript
+
+    pushd $REPOFOLDER
+
+    # because `--add-file` is not available except in very recent versions of `git`, manually append binaries
+    git archive --format=tar -o $DOCKERFOLDER/gravity.tar --prefix=gravity/ HEAD
```
Comment (dzmitryhil) [inline tests/build-container.sh:42]: What is the purpose to have gravity.tar.gz and gravity.tar at the same time?

**Review by dzmitryhil — CHANGES_REQUESTED** [review]

File: tests/all-up-test.sh [hunk 5]
```diff
@@ -25,4 +25,4 @@ ALCHEMY_ID=$2
 set -u
```
Comment (dzmitryhil) [inline tests/all-up-test.sh:25]: Also, it would be nice to add a new script "[build-test-artifacts.sh]" which rebuilds the local artifacts.

**Review by dzmitryhil — CHANGES_REQUESTED** [review]

File: tests/build-container.sh (hunk 4 above)
Comment (AaronKutch) [reply tests/build-container.sh:42]: It says in the comment. I guess I should reword it to `because  --add-file is not available except in very recent versions of git, we cannot add them through the archive command and need to add them to the tar file ourselves, before it is compressed`

File: tests/build-container.sh (hunk 1 above)
Comment (AaronKutch) [reply tests/build-container.sh:29]: ok

File: tests/container-scripts/integration-tests.sh (hunk 3 above)
Comment (AaronKutch) [reply tests/container-scripts/integration-tests.sh:21]: ok

File: tests/build-container.sh (hunk 2 above)
Comment (AaronKutch) [reply tests/build-container.sh:34]: I tried updating my npm version and broke `npm install` locally (so now I can only build it in the container), so I'm deferring using the local npm artifacts at least until leviathan finishes his npm fixes and I have the time to fix it.

File: tests/container-scripts/integration-tests.sh (hunk 3 above)
Comment (AaronKutch) [reply tests/container-scripts/integration-tests.sh:21]: I think you may have intended for me to pass the RUN_ARGS between scripts, but for now I am only setting it using USE_LOCAL_ARTIFACTS wherever it is needed. I found that passing it between scripts adds too much complexity and is not worth the hassle.

File: tests/build-container.sh (hunk 4 above)
Comment (dzmitryhil) [reply tests/build-container.sh:38]: so maybe we could you "tar" only?

Comment (AaronKutch) [reply tests/build-container.sh:38]: I suppose so since it is uncompressed when adding to the dockerfile

Comment (dzmitryhil) [reply tests/build-container.sh:40]: let's use "tar" only then

File: tests/build-container.sh (hunk 2 above)
Comment (dzmitryhil) [reply tests/build-container.sh:31]: We can add this improvement later, so no we go without it, and then when all fixes are done with it :-) Anyway, the update of the artifact should affect the build.
